### PR TITLE
Initial Attempt at Crashlytics Beta Initialization

### PR DESCRIPTION
**Pull request by ohayon**
This PR adds a feature-gated (by the 'FASTLANE_ENABLE_CRASHLYTICS_BETA_INITIALIZATION' env var) command, `beta_init` (open to changing the name of this). 

This will detect your project, parse your API Key and Secret from the RSBP or prompt the user for them to be entered and generate the MVF (Minimum Viable Fastfile) or lane to be copy and pasted if a Fastfile already exists.

### CODE REVIEW

File: fastlane/lib/fastlane.rb [hunk 1]
```diff
@@ -19,6 +20,7 @@
 require 'fastlane/documentation/docs_generator'
 require 'fastlane/other_action'
 require 'fastlane/plugins/plugins'
+require 'fastlane/setup/crashlytics_beta'
```
Comment (KrauseFx) [inline fastlane/lib/fastlane.rb]: I'd move that to https://github.com/fastlane/fastlane/pull/5201/files#diff-87e6d7e2cbe2a0420c4c4efb74cd3cbcR62, as long as it's a feature flag at least 👍

File: fastlane/lib/fastlane/setup/crashlytics_beta.rb [hunk 2]
```diff
@@ -0,0 +1,88 @@
+module Fastlane
+  class CrashlyticsBeta
+    def run
+      config = {}
+      FastlaneCore::Project.detect_projects(config)
+      project = FastlaneCore::Project.new(config)
+      keys = keys_from_project(project)
+
+      if File.exist?('fastlane/Fastfile') || File.exist?('Fastfile')
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/setup/crashlytics_beta.rb]: Check out `FastlaneFolder` for how to detect a `Fastfile` properly 👍

File: fastlane/lib/fastlane/setup/crashlytics_beta.rb [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        UI.header('Copy and Paste the following lane into your Fastfile to use Crashlytics Beta!')
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/setup/crashlytics_beta.rb]: paste should be lower case, right?

File: fastlane/lib/fastlane/setup/crashlytics_beta.rb [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        puts lane_template(keys[:api_key], keys[:build_secret], project.schemes.first)
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/setup/crashlytics_beta.rb]: Maybe this should be the same color as `UI.command` or `UI.command_output`

File: fastlane/lib/fastlane/setup/crashlytics_beta.rb [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+      else
+        fastfile = fastfile_template(keys[:api_key], keys[:build_secret], project.schemes.first)
+        FileUtils.mkdir_p('fastlane')
+        File.open('fastlane/Fastfile', 'w') { |f| f.write(fastfile) }
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/setup/crashlytics_beta.rb]: You can just use `File.write(path, content)`

File: fastlane/lib/fastlane/setup/crashlytics_beta.rb [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      end
+    end
+
+    def keys_from_project(project)
+      target_name = project.default_build_settings(key: 'TARGETNAME')
+      require 'xcodeproj'
+      path = project.is_workspace ? project.path.gsub('xcworkspace', 'xcodeproj') : project.path
+      UI.crash! "No project available at path #{path}" unless File.exist?(path)
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/setup/crashlytics_beta.rb]: Use `()` around the crash text

File: fastlane/lib/fastlane/setup/crashlytics_beta.rb [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+      xcode_project = Xcodeproj::Project.open(path)
+      target = xcode_project.targets.find { |t| t.name == target_name }
+      scripts = target.build_phases.select { |t| t.class == Xcodeproj::Project::Object::PBXShellScriptBuildPhase }
+      crash_script = scripts.find { |s| includes_run_script?(s.shell_script) }
+      UI.user_error! "Unable to find Crashlytics Run Script Build Phase" if crash_script.nil?
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/setup/crashlytics_beta.rb]: Same here https://github.com/fastlane/fastlane/pull/5201/files#r68309742

File: fastlane/lib/fastlane/setup/crashlytics_beta.rb [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+      else
+        fastfile = fastfile_template(keys[:api_key], keys[:build_secret], project.schemes.first)
+        FileUtils.mkdir_p('fastlane')
+        File.open('fastlane/Fastfile', 'w') { |f| f.write(fastfile) }
+      end
+    end
+
+    def keys_from_project(project)
+      target_name = project.default_build_settings(key: 'TARGETNAME')
+      require 'xcodeproj'
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/setup/crashlytics_beta.rb]: Why do we require on the second line of the method?

File: fastlane/lib/fastlane/setup/crashlytics_beta.rb (hunk 4 above)
Comment (ohwutup) [inline fastlane/lib/fastlane/setup/crashlytics_beta.rb]: does one of those options remove the time stamp?

File: fastlane/lib/fastlane.rb (hunk 1 above)
Comment (ohwutup) [inline fastlane/lib/fastlane.rb]: good point!

File: fastlane/lib/fastlane/setup/crashlytics_beta.rb [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+          keys[:build_secret] = UI.input('Build Secret:')
+          break if build_secret_valid?(keys[:build_secret])
+          UI.important "Invalid Build Secret, Please Try Again!"
+        end
+        keys
+      end
+    end
+
+    def api_key_valid?(key)
+      key.length == 40
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/setup/crashlytics_beta.rb]: Just to be sure I'd change that to `key.to_s.length == 50`, so it doesn't crash when it's `nil`.

**Comment by asfalcone** [comment]
Should we do anything to specify that this command is for iOS projects only right now? Like in the check for a .xcodeproj add a warning that says this is only for iOS projects until we implement it for android?

File: fastlane/lib/fastlane/commands_generator.rb [hunk 10]
```diff
@@ -58,7 +58,12 @@ def run
         c.description = 'Helps you with your initial fastlane setup'
 
         c.action do |args, options|
-          Fastlane::Setup.new.run
+          if args[0] == 'beta' && FastlaneCore::Feature.enabled?('FASTLANE_ENABLE_CRASHLYTICS_BETA_INITIALIZATION')
```
Comment (asfalcone) [inline fastlane/lib/fastlane/commands_generator.rb]: Is the command that developers will use `init beta` or `beta_init`? I'm a little confused by this.

Comment (ohwutup) [inline fastlane/lib/fastlane/commands_generator.rb]: `init beta` (for now at least. definitely open to suggestions!)

Comment (asfalcone) [inline fastlane/lib/fastlane/commands_generator.rb]: cool, that makes sense, then should [this line](https://github.com/fastlane/fastlane/pull/5201/files/14ea820d33ee8d4bc18bb5680b518fe3f6ea5221#diff-6d9cdfc0945e56dd3efea6ab7e8d11aeR2) be updated to reflect that?

File: fastlane/lib/fastlane/setup/crashlytics_beta.rb [hunk 11]
```diff
@@ -1,6 +1,7 @@
 module Fastlane
   class CrashlyticsBeta
     def run
+      UI.user_error!('Beta by Crashlytics configuration is currently only available for iOS projects.') unless Setup.new.is_ios?
```
Comment (asfalcone) [inline fastlane/lib/fastlane/setup/crashlytics_beta.rb]: Cool. Definitely better to let folks know early 👍

**Comment by asfalcone** [comment]
Finished my pass. Just 1 more comment on there :)

**Comment by asfalcone** [comment]
Looks good to me 🚀

**Comment by KrauseFx** [comment]
I understand this is a feature flag and a prototype with a tight deadline, but we shouldn't forget to add some tests for this once we go ahead 🚀

File: fastlane/lib/fastlane/setup/crashlytics_beta.rb [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+      project = FastlaneCore::Project.new(config)
+      keys = keys_from_project(project)
+
+      if FastlaneFolder.setup?
+        UI.header('Copy and paste the following lane into your Fastfile to use Crashlytics Beta!')
+        puts lane_template(keys[:api_key], keys[:build_secret], project.schemes.first).cyan
+      else
+        fastfile = fastfile_template(keys[:api_key], keys[:build_secret], project.schemes.first)
+        FileUtils.mkdir_p('fastlane')
+        File.write('fastlane/Fastfile', fastfile)
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/setup/crashlytics_beta.rb]: I didn't run this myself, but we should probably print out some information to the user to let them know we created this file for them, right?

Comment (ohwutup) [inline fastlane/lib/fastlane/setup/crashlytics_beta.rb]: Yeah definitely, @mpirri and I had discussed this and we definitely plan on doing that.

**Comment by ohwutup** [comment]
Yeah, definitely going to write tests for this also, we had had that as a checklist item on our trello card but then nixed it for this week because we wanted to get it out. Made a new card for next week here: https://trello.com/c/n6XBthNV/368-init-beta-work

**Pull request merged**